### DISCUSSION TITLE
OPS for QM/MM

### DISCUSSION THREAD

**Discussion by davidkastner** [openpathsampling/openpathsampling#1006] [Q&A] [answered]
Could path sampling of QM/MM simulations be implemented with OpenPathSampling? For example, in our lab we use Amber, OpenMM, and TeraChem to run QM/MM simulations but are looking for the best way to implement path sampling. OpenPathSampling looks like the best tool available but haven't seen any discussions of its use with QM/MM simulations.

**Comment by dwhswenson** [comment, marked as answer]
In principle, absolutely yes. In practice this may require implementing a new `DynamicsEngine`, which is the interface between OpenPathSampling and the underlying MD engine. This is something that we'd be very interested in having contributions for, and I would be happy to help with the details.

One of the potential challenges comes from the fact that each snapshot along a trajectory formally needs to be a restart point. That is to say, it should include all the information required to restart the simulation. I'm not sure what's required for various QM/MM methods; however, in Car-Parrinello MD (or even regular MD with a Nosé-Hoover type thermostat) this technically should include the extended phase space variables -- for a restart, you need these (in addition to positions/velocities). This information can be added to OPS as "snapshot features."

In terms of how to go about this, the documentation on adding engines is (like much of our documentation) incomplete, but it happens to be one of the more solid parts of the developer docs, because this is an area where a user with expertise in the specific engine can rather easily contribute to OPS. You can find that documentation here: http://openpathsampling.org/latest/developers/engines.html. And I'd be happy to help with any problems along the way and to help with getting code ready for inclusion in a release.

**Reply by davidkastner** [reply]
Thank you for the thorough reply. It sounds quite promising. I would be very interested in working on implementing this.
